core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -297,7 +296,7 @@ public void add(T input) throws IOException {
      * @return a list of inputs for successful operations
      * @throws IOException if a non-Alluxio error occurs
      */
-    public List<T> getResult() throws IOException {
+    public synchronized List<T> getResult() throws IOException {
```

Comment:
think adding synchronized on both getResult and add func will make their access serialized since the they are locking on the same monitor of the object. even though u r not using both func concurrently in this case, since getResult is blockingly wait for result, it will prevent add() being called. I don't think this is best practice to add synchronized to make it threadsafe. I would suggest to make mBatchesResult a arrayblockingq and use poll to individually wait for result in sequence. and take a lock on modification on those collections in submitBatch and add

Reply:
I don't think it would be easy to modify it to a blocking queue. I think it is acceptable to add synchronized for now, since the serialized access part is an in-memory operation. We can discuss how to modify it to a blocking queue in a later PR.

Reply:
It is possible to introduce a hang in the system which could bring system downtime. Please reconsider the suggestion.

Reply:
Why do you think It is possible to introduce a hang? I don't see a problem with this implementation. @ssz1997

Reply:
@qian0817  in your implementation it wouldn't hang, bcos u did call all the add() calls before you do getresult(), but this class is not ThreadSafe given the situation i described:
if getresult() is called before anyother concurrent add() gets called, add() will be blocked until getresult() got his first result from the list. Its like producer couldn't produce to the q jus bcos consumer is blockingly consuming the front of the q. 
So if other people uses this class in this case, this is a possible hang case.

Reply:
ok cool, a comment would be good enough, thanks